lib/ui/painting/image_decoder_impeller.cc

```diff
@@ -490,17 +490,28 @@ void ImageDecoderImpeller::Decode(fml::RefPtr<ImageDescriptor> descriptor,
 
         sk_sp<DlImage> image;
         std::string decode_error;
-        if (context->GetCapabilities()->SupportsBufferToTextureBlits()) {
-          std::tie(image, decode_error) = UploadTextureToPrivate(
-              context, bitmap_result.device_buffer, bitmap_result.image_info,
-              bitmap_result.sk_bitmap, gpu_disabled_switch);
-          result(image, decode_error);
+        auto upload_texture_and_invoke_result = [result, context, bitmap_result,
+                                                 gpu_disabled_switch]() {
+          sk_sp<DlImage> image;
+          std::string decode_error;
+          if (context->GetCapabilities()->SupportsBufferToTextureBlits()) {
+            std::tie(image, decode_error) = UploadTextureToPrivate(
+                context, bitmap_result.device_buffer, bitmap_result.image_info,
+                bitmap_result.sk_bitmap, gpu_disabled_switch);
+            result(image, decode_error);
+          } else {
+            std::tie(image, decode_error) = UploadTextureToStorage(
+                context, bitmap_result.sk_bitmap, gpu_disabled_switch,
+                impeller::StorageMode::kDevicePrivate,
+                /*create_mips=*/true);
+            result(image, decode_error);
+          }
+        };
+        if (context->GetBackendType() ==
+            impeller::Context::BackendType::kOpenGLES) {
+          io_runner->PostTask(upload_texture_and_invoke_result);
```

Comment:
Hmm, I would have assumed that in the backend implementation, `UploadTextureToFoo` would rethread to a worker on a thread that returns `true` for `CanReactorReactOnCurrentThreadNow`. Its unfortunate we have to care about the backend here. Can we file an issue to track that that down? This is fine though.

Reply:
Done: https://github.com/flutter/flutter/issues/147520